Source/WebCore/rendering/svg/SVGResourcesCycleSolver.cpp

```diff
@@ -50,134 +41,75 @@ SVGResourcesCycleSolver::SVGResourcesCycleSolver(RenderElement& renderer, SVGRes
 
 SVGResourcesCycleSolver::~SVGResourcesCycleSolver() = default;
 
-bool SVGResourcesCycleSolver::resourceContainsCycles(RenderElement& renderer) const
-{
-    LOG_DEBUG_CYCLE("\n(%p) Check for cycles\n", &renderer);
-
-    // First operate on the resources of the given renderer.
-    // <marker id="a"> <path marker-start="url(#b)"/> ...
-    // <marker id="b" marker-start="url(#a)"/>
-    if (auto* resources = SVGResourcesCache::cachedResourcesForRenderer(renderer)) {
-        WeakHashSet<RenderSVGResourceContainer> resourceSet;
-        resources->buildSetOfResources(resourceSet);
-
-        LOG_DEBUG_CYCLE("(%p) Examine our cached resources\n", &renderer);
-
-        // Walk all resources and check whether they reference any resource contained in the resources set.
-        for (auto& resource : resourceSet) {
-            LOG_DEBUG_CYCLE("(%p) Check %p\n", &renderer, &resource);
-            if (m_allResources.contains(resource))
-                return true;
-
-            // Now check if the resources themselves contain cycles.
-            if (resourceContainsCycles(resource))
-                return true;
-        }
+class ActiveResourceAcquisition {
+public:
+    ActiveResourceAcquisition(WeakHashSet<RenderSVGResourceContainer>& set, RenderSVGResourceContainer& resource)
+        : m_set(set)
+        , m_resource(resource)
+    {
+        m_set.add(m_resource);
     }
 
-    LOG_DEBUG_CYCLE("(%p) Now the children renderers\n", &renderer);
-
-    // Then operate on the child resources of the given renderer.
-    // <marker id="a"> <path marker-start="url(#b)"/> ...
-    // <marker id="b"> <path marker-start="url(#a)"/> ...
-    for (auto& child : childrenOfType<RenderElement>(renderer)) {
-
-        LOG_DEBUG_CYCLE("(%p) Checking child %p\n", &renderer, &child);
+    ~ActiveResourceAcquisition()
+    {
+        m_set.remove(m_resource);
+    }
 
-        if (auto* childResources = SVGResourcesCache::cachedResourcesForRenderer(child)) {
+private:
+    WeakHashSet<RenderSVGResourceContainer>& m_set;
+    RenderSVGResourceContainer& m_resource;
+};
 
-            LOG_DEBUG_CYCLE("(%p) Child %p had cached resources. Check them.\n", &renderer, &child);
+bool SVGResourcesCycleSolver::resourceContainsCycles(RenderSVGResourceContainer& resource)
+{
+    if (m_acyclicResources.contains(resource))
```

Comment:
Why do we check the `m_acyclicResources` only? And what is the definition of `acyclicResources` here since we `breakCycle()` for the cyclic ones. I think we should check `allVisitedResources` instead.

Reply:
This is an early return for when we had already determined that resource doesn't have a cycle.
Without this check, we can end up checking the same resource doesn't have a cycle over & over.